File: src/main/java/burp/StorageManager.java
package burp;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.reflect.TypeToken;

import java.io.*;
import java.lang.reflect.Type;
import java.util.ArrayList;
import java.util.List;
import java.util.Set;

public class StorageManager {
    private final File storageFile;
    private final File configFile;
    private final Gson gson;
    private final PrintWriter stderr;

    public StorageManager(String storagePath, String configPath, PrintWriter stderr) {
        this.storageFile = new File(storagePath);
        this.configFile = new File(configPath);
        this.stderr = stderr;
        this.gson = new GsonBuilder().setPrettyPrinting().create();
    }

    public synchronized void saveAttack(AttackEntry entry) {
        List<AttackEntry> attacks = loadAttacks();
        attacks.add(entry);
        saveAll(attacks);
        cachedAttacks = attacks;
        lastLoadedTime = storageFile.lastModified();
    }

    private long lastLoadedTime = 0;
    private List<AttackEntry> cachedAttacks = null;

    public synchronized List<AttackEntry> loadAttacks() {
        if (!storageFile.exists()) {
            cachedAttacks = new ArrayList<>();
            return cachedAttacks;
        }

        long lastModified = storageFile.lastModified();
        if (cachedAttacks != null && lastModified <= lastLoadedTime) {
            return new ArrayList<>(cachedAttacks); // Return copy to prevent external modification affecting cache
        }

        try (Reader reader = new FileReader(storageFile)) {
            Type listType = new TypeToken<ArrayList<AttackEntry>>(){}.getType();
            List<AttackEntry> attacks = gson.fromJson(reader, listType);
            if (attacks != null) {
                boolean modified = false;
                for (AttackEntry attack : attacks) {
                    if (attack.getId() == null || attack.getId().isEmpty()) {
                        attack.ensureId();
                        modified = true;
                    }
                }
                if (modified) {
                    saveAll(attacks);
                    // saveAll updates the file, so lastModified will change.
                    // We should update lastLoadedTime to the new lastModified.
                    lastModified = storageFile.lastModified();
                }
                
                cachedAttacks = attacks;
                lastLoadedTime = lastModified;
                return new ArrayList<>(cachedAttacks);
            }
            cachedAttacks = new ArrayList<>();
            return cachedAttacks;
        } catch (IOException e) {
            stderr.println("Error loading attacks: " + e.getMessage());
            e.printStackTrace(stderr);
            return new ArrayList<>();
        }
    }

    public synchronized void overwriteAttacks(List<AttackEntry> attacks) {
        saveAll(attacks);
        cachedAttacks = new ArrayList<>(attacks);
        lastLoadedTime = storageFile.lastModified();
    }

    public synchronized List<AttackEntry> deleteAttacks(Set<String> ids, boolean permanent) {
        List<AttackEntry> allAttacks = loadAttacks();
        boolean changed = false;

        if (permanent) {
            int initialSize = allAttacks.size();
            allAttacks.removeIf(a -> ids.contains(a.getId()));
            if (allAttacks.size() < initialSize) changed = true;
        } else {
            for (AttackEntry attack : allAttacks) {
                if (ids.contains(attack.getId())) {
                    attack.setDeleted(true);
                    changed = true;
                }
            }
        }

        if (changed) {
            saveAll(allAttacks);
        }
        return allAttacks;
    }

    public synchronized List<AttackEntry> restoreAttacks(Set<String> ids) {
        List<AttackEntry> allAttacks = loadAttacks();
        boolean changed = false;

        for (AttackEntry attack : allAttacks) {
            if (ids.contains(attack.getId())) {
                attack.setDeleted(false);
                changed = true;
            }
        }

        if (changed) {
            saveAll(allAttacks);
        }
        return allAttacks;
    }

    public synchronized List<AttackEntry> emptyTrash() {
        List<AttackEntry> allAttacks = loadAttacks();
        int initialSize = allAttacks.size();
        allAttacks.removeIf(AttackEntry::isDeleted);
        
        if (allAttacks.size() < initialSize) {
            saveAll(allAttacks);
        }
        return allAttacks;
    }

    private void saveAll(List<AttackEntry> attacks) {
        try (Writer writer = new FileWriter(storageFile)) {
            gson.toJson(attacks, writer);
            lastLoadedTime = storageFile.lastModified();
            cachedAttacks = new ArrayList<>(attacks); // Update cache
        } catch (IOException e) {
            // stderr.println("Error saving attacks: " + e.getMessage());
            // e.printStackTrace(stderr);
        }
    }

    public synchronized void saveConfig(ExtensionConfig config) {
        try (Writer writer = new FileWriter(configFile)) {
            gson.toJson(config, writer);
        } catch (IOException e) {
            stderr.println("Error saving config: " + e.getMessage());
            e.printStackTrace(stderr);
        }
    }

    public synchronized ExtensionConfig loadConfig() {
        if (!configFile.exists()) {
            return new ExtensionConfig(); // Return defaults
        }

        try (Reader reader = new FileReader(configFile)) {
            ExtensionConfig config = gson.fromJson(reader, ExtensionConfig.class);
            return config != null ? config : new ExtensionConfig();
        } catch (IOException e) {
            stderr.println("Error loading config: " + e.getMessage());
            e.printStackTrace(stderr);
            return new ExtensionConfig();
        }
    }
    public synchronized void exportAttacks(File destination) throws IOException {
        List<AttackEntry> attacks = loadAttacks();
        try (Writer writer = new FileWriter(destination)) {
            gson.toJson(attacks, writer);
        }
    }

    public synchronized void importAttacks(File source) throws IOException {
        if (!source.exists()) {
            throw new FileNotFoundException("Import file not found: " + source.getAbsolutePath());
        }

        try (Reader reader = new FileReader(source)) {
            Type listType = new TypeToken<ArrayList<AttackEntry>>(){}.getType();
            List<AttackEntry> importedAttacks = gson.fromJson(reader, listType);
            
            if (importedAttacks != null && !importedAttacks.isEmpty()) {
                List<AttackEntry> currentAttacks = loadAttacks();
                
                // Collect existing IDs to prevent duplicates
                java.util.Set<String> existingIds = new java.util.HashSet<>();
                for (AttackEntry entry : currentAttacks) {
                    if (entry.getId() != null) {
                        existingIds.add(entry.getId());
                    }
                }
                
                // Filter and add only unique imported entries
                List<AttackEntry> uniqueNewAttacks = new java.util.ArrayList<>();
                for (AttackEntry imported : importedAttacks) {
                    // Ensure the entry has an ID
                    if (imported.getId() == null || imported.getId().isEmpty()) {
                        imported.ensureId();
                    }
                    // Only add if ID doesn't already exist
                    if (!existingIds.contains(imported.getId())) {
                        uniqueNewAttacks.add(imported);
                        existingIds.add(imported.getId()); // Track to prevent duplicates within import file
                    }
                }
                
                if (!uniqueNewAttacks.isEmpty()) {
                    currentAttacks.addAll(uniqueNewAttacks);
                    saveAll(currentAttacks);
                }
            }
        }
    }
}

File: src/main/java/com/projectkepler/burp/StorageManager.kt
package com.projectkepler.burp

import burp.api.montoya.logging.Logging
import com.google.gson.GsonBuilder
import com.google.gson.reflect.TypeToken
import java.io.*
import java.lang.reflect.Type

/**
 * Manages persistence of attack data and configuration.
 * Why: Centralizing storage logic ensures thread-safe access to the JSON files
 * and provides a caching layer to minimize disk I/O during UI refreshes.
 */
class StorageManager(
    storagePath: String,
    configPath: String,
    private val logging: Logging?
) {
    private val storageFile = File(storagePath)
    private val configFile = File(configPath)
    private val gson = GsonBuilder().setPrettyPrinting().create()

    private var lastLoadedTime: Long = 0
    private var cachedAttacks: MutableList<AttackEntry>? = null

    @Synchronized
    fun saveAttack(entry: AttackEntry) {
        val attacks = loadAttacks()
        attacks.add(entry)
        saveAll(attacks)
    }

    @Synchronized

    fun updateAttack(entry: AttackEntry) {
        val attacks = loadAttacks()
        val index = attacks.indexOfFirst { it.id == entry.id }
        if (index != -1) {
            attacks[index] = entry
            saveAll(attacks)
        }
    }

    /**
     * Loads attacks from disk with a caching strategy based on file modification time.
     * Why: Burp UI can trigger multiple refreshes; caching prevents redundant parsing of potentially large JSON files.
     */
    @Synchronized
    fun loadAttacks(): MutableList<AttackEntry> {
        if (!storageFile.exists()) {
            return mutableListOf<AttackEntry>().also { cachedAttacks = it }
        }

        val lastModified = storageFile.lastModified()
        cachedAttacks?.let {
            if (lastModified <= lastLoadedTime) {
                return it.toMutableList()
            }
        }

        return try {
            FileReader(storageFile).use { reader ->
                val mapListType: Type = object : TypeToken<ArrayList<Map<String, Any?>>>() {}.type
                val attackMaps: MutableList<Map<String, Any?>>? = gson.fromJson(reader, mapListType)

                var modified = false
                attackMaps?.forEach { map ->
                    @Suppress("UNCHECKED_CAST")
                    val mutableMap = map as? MutableMap<String, Any?> ?: return@forEach
                    if (mutableMap["id"] == null || (mutableMap["id"] as? String).isNullOrBlank()) {
                        mutableMap["id"] = java.util.UUID.randomUUID().toString()
                        modified = true
                    }
                }

                val attacks: MutableList<AttackEntry> = if (attackMaps != null) {
                    val attacksJson = gson.toJson(attackMaps)
                    val attackListType: Type = object : TypeToken<ArrayList<AttackEntry>>() {}.type
                    gson.fromJson<ArrayList<AttackEntry>>(attacksJson, attackListType) ?: mutableListOf()
                } else {
                    mutableListOf()
                }

                if (modified) {
                    saveAll(attacks)
                }

                attacks.also { attackList ->
                    cachedAttacks = attackList.toMutableList()
                    lastLoadedTime = storageFile.lastModified()
                }
            }
        } catch (e: IOException) {
            logging?.logToError("Error loading attacks: ${e.message}")
            mutableListOf()
        }
    }

    @Synchronized
    fun deleteAttacks(ids: Set<String>, permanent: Boolean): List<AttackEntry> {
        val allAttacks = loadAttacks()
        var changed = false

        if (permanent) {
            changed = allAttacks.removeIf { it.id in ids }
        } else {
            allAttacks.forEach { attack ->
                if (attack.id in ids && !attack.deleted) {
                    attack.deleted = true
                    changed = true
                }
            }
        }

        if (changed) saveAll(allAttacks)
        return allAttacks
    }

    @Synchronized
    fun restoreAttacks(ids: Set<String>): List<AttackEntry> {
        val allAttacks = loadAttacks()
        var changed = false

        allAttacks.forEach { attack ->
            if (attack.id in ids && attack.deleted) {
                attack.deleted = false
                changed = true
            }
        }

        if (changed) saveAll(allAttacks)
        return allAttacks
    }

    @Synchronized
    fun emptyTrash(): List<AttackEntry> {
        val allAttacks = loadAttacks()
        if (allAttacks.removeIf { it.deleted }) {
            saveAll(allAttacks)
        }
        return allAttacks
    }

    private fun saveAll(attacks: List<AttackEntry>) {
        try {
            FileWriter(storageFile).use { writer ->
                gson.toJson(attacks, writer)
                lastLoadedTime = storageFile.lastModified()
                cachedAttacks = attacks.toMutableList()
            }
        } catch (e: IOException) {
            logging?.logToError("Error saving attacks: ${e.message}")
        }
    }

    @Synchronized
    fun saveConfig(config: ExtensionConfig) {
        try {
            FileWriter(configFile).use { writer -> gson.toJson(config, writer) }
        } catch (e: IOException) {
            logging?.logToError("Error saving config: ${e.message}")
        }
    }

    @Synchronized
    fun loadConfig(): ExtensionConfig {
        if (!configFile.exists()) return ExtensionConfig()

        return try {
            FileReader(configFile).use { reader ->
                gson.fromJson(reader, ExtensionConfig::class.java) ?: ExtensionConfig()
            }
        } catch (e: IOException) {
            logging?.logToError("Error loading config: ${e.message}")
            ExtensionConfig()
        }
    }

    @Synchronized
    @Throws(IOException::class)
    fun exportAttacks(destination: File) {
        val attacks = loadAttacks()
        FileWriter(destination).use { writer -> gson.toJson(attacks, writer) }
    }

    @Synchronized
    @Throws(IOException::class)
    fun importAttacks(source: File) {
        if (!source.exists()) throw FileNotFoundException("Import file not found: ${source.absolutePath}")

        FileReader(source).use { reader ->
            val mapListType: Type = object : TypeToken<ArrayList<Map<String, Any?>>>() {}.type
            val attackMaps: MutableList<Map<String, Any?>>? = gson.fromJson(reader, mapListType)

            attackMaps?.forEach { map ->
                @Suppress("UNCHECKED_CAST")
                val mutableMap = map as? MutableMap<String, Any?> ?: return@forEach
                if (mutableMap["id"] == null || (mutableMap["id"] as? String).isNullOrBlank()) {
                    mutableMap["id"] = java.util.UUID.randomUUID().toString()
                }
            }

            val importedAttacks: List<AttackEntry> = if (attackMaps != null) {
                val attacksJson = gson.toJson(attackMaps)
                val attackListType: Type = object : TypeToken<ArrayList<AttackEntry>>() {}.type
                gson.fromJson<ArrayList<AttackEntry>>(attacksJson, attackListType) ?: emptyList()
            } else {
                emptyList()
            }

            if (importedAttacks.isNotEmpty()) {
                val currentAttacks = loadAttacks()
                val existingIds = currentAttacks.map { entry -> entry.id }.toSet()

                val uniqueNewAttacks = importedAttacks.filter { entry -> entry.id !in existingIds }

                if (uniqueNewAttacks.isNotEmpty()) {
                    currentAttacks.addAll(uniqueNewAttacks)
                    saveAll(currentAttacks)
                }
            }
        }
    }
}
